File: 02_common_uses/04_interfaces.cpp
#include <iostream>
#include <concepts>
class DrawableInterface {
public:
    virtual void Draw() = 0;
};


class Circle final: public DrawableInterface {
    public:
    void Draw() final {
        std::cout<<"circle"<<std::endl;
    }
};

class Square final: public DrawableInterface {
    public:
    void Draw() final {
        std::cout<<"square"<<std::endl;
    }
};

void RuntimePolymorphism(DrawableInterface& drawable){
    drawable.Draw();
}

template <typename T>
concept Drawable = requires(T& obj) {
    { obj.Draw() } -> std::same_as<void>; // Requires a draw() method that returns void
};

template <Drawable T> void CompileTimePolymorphism(T& drawable){
    drawable.Draw();
}
int main(){
    Circle c {};
    RuntimePolymorphism(c);
    CompileTimePolymorphism(c);
    return 0;
}

File: 02_common_uses/04_interfaces.rs
trait Drawable{
    fn draw(&self);
}

struct Circle {

}

struct Square{

}

impl Drawable for Circle {
    fn draw(&self){
        println!("circle");
    }
}

impl Drawable for Square {
    fn draw(&self){
        println!("square");
    }
}

fn runtime_polymorphism(drawable: &Drawable){
    drawable.draw();
}

fn compile_time_polymorphism<T: Drawable>( drawable: &T){
    drawable.draw();
}



fn main(){
    let c = Circle{};
    runtime_polymorphism(c);
    compile_time_polymorphism(c);

}
